repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeFilterLoader.scala

```diff
@@ -0,0 +1,321 @@
+    logger.info("{} -- Application collection has been modified. Application that changed: {}", nodeId, e.payload)
+    val duplicates = e.payload().asScala.groupBy(identity).collect { case (x, List(_, _, _*)) => x }
+    if (duplicates.isEmpty) {
+      healthCheckServiceProxy.resolveIssue(ApplicationDeploymentHealthReport)
+    } else {
+      val message = s"Please review your artifacts directory, multiple versions of the artifact${if (duplicates.size > 1) "s" else ""} exist! (${duplicates.mkString(",")})"
+      healthCheckServiceProxy.reportIssue(ApplicationDeploymentHealthReport, message, Severity.BROKEN)
+      logger.error(message)
+    }
+    if (artifactManager.allArtifactsLoaded) {
+      configurationCheck()
+    }
+  }
+
+  def setServletContext(newServletContext: ServletContext): Unit = {
+    currentServletContext = Option(newServletContext)
+    configurationCheck()
+  }
+
+  def getFilterContextList: Option[FilterContextList] = {
+    currentFilterContextRegistrar.map(_.bind())
+  }
+
+  def getTracingHeaderConfig: Option[TracingHeaderConfig] = currentSystemModel.map(_.getTracingHeader)
+
+  /**
+    * Triggered each time the event service triggers an app deploy and when the system model is updated.
+    */
+  private def configurationCheck(): Unit = {
+    (currentSystemModel, currentServletContext, artifactManager.allArtifactsLoaded) match {
+      case (Some(systemModel), Some(servletContext), true) =>
+        updateConfiguration(systemModel, servletContext)
+      case _ =>
+        logger.trace("{} -- Not ready to update yet.", nodeId)
+    }
+  }
+
+  private def updateConfiguration(systemModel: SystemModel, servletContext: ServletContext): Unit = {
+    configurationLock.synchronized {
+      val interrogator = new SystemModelInterrogator(nodeId)
+
+      Option(interrogator.getLocalCluster(systemModel).orElse(null)) match {
+        case Some(localCluster) =>
+          healthCheckServiceProxy.resolveIssue(SystemModelConfigHealthReport)
+          try {
+            // Only if we've been configured with some filters should we get a new list
+            // Sometimes we won't have any filters
+            val newFilterChain = Option(localCluster.getFilters) match {
+              case Some(listOfFilters: FilterList) => buildFilterContexts(servletContext, listOfFilters.getFilter.asScala.toList)
+              case _ => List.empty[FilterContext]
+            }
+
+            // Set the new FilterContextRegistrar and
+            // set the Close flag on the old one.
+            currentFilterContextRegistrar.synchronized {
+              val previousFilterContextRegistrar = currentFilterContextRegistrar
+              currentFilterContextRegistrar = Option(new FilterContextRegistrar(newFilterChain, Option(localCluster.getFilters.getBypassUriRegex)))
+              previousFilterContextRegistrar
+            }.foreach(_.close())
+
+            if (logger.underlying.isDebugEnabled) {
```

Comment:
Cool. So that alleviates any concerns I had about the performance hit.

Reply:
Although this still causes the iteration over all of the filters in the FilterChain regardless of if we are actually going to log them.
However, since this is only done on a configuration update, I'm not overly concerned either way.

Reply:
Yeah, I wouldn't be concerned about it... but if you are, you could use a stream and log4j's support for lazy evaluation vs lamdas. Something like:
```
val filterChainInfo = newFilterChain.toStream.map(ctx => ctx.filter.getClass.getName)
logger.debug("{} -- Repose filter chain: {}", () -> nodeId, () -> filterChainInfo.mkString(","))
```

That is a weird mix of Java and Scala, but you get the idea. The lambdas wouldn't even be necessary is the Scala Logging library took arguments by name, but alas, that does not appear to be the case.

Reply:
After thinking about this over the last few days, I don't think I'm super concerned anymore.